### PR TITLE
[DO NOT SQUASH] LLVM upstream merge

### PR DESCRIPTION
**Pull request by giuseros**
This is the LLVM upstream merge for this release.

### CODE REVIEW

File: mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp [hunk 1]
```diff
@@ -620,8 +622,8 @@ class DivConverter final : public OpConversionPattern<migraphx::DivOp> {
   matchAndRewrite(migraphx::DivOp op, OpAdaptor adaptor,
                   ConversionPatternRewriter &rewriter) const final {
     Location loc = op.getLoc();
-    TypedValue<TensorType> inATensor = op.getInA();
-    TypedValue<TensorType> inBTensor = op.getInB();
+    auto inATensor = cast<TypedValue<TensorType>>(op.getInA());
```
Comment (krzysz00) [inline mlir/lib/Conversion/MIGraphXToTosa/MIGraphXToTosa.cpp]: Did they revert the automatic `typedValue` thing? Or is this just something with how tensors are defined upstream now?

File: mlir/lib/Dialect/Rock/IR/RockDialect.cpp [hunk 2]
```diff
@@ -697,7 +697,8 @@ static LogicalResult checkGemmSize(Value matrix, Operation *op,
   while (auto transform = raw.getDefiningOp<rock::TransformOp>())
     raw = transform.getInput();
   ShapedType type = raw.getType().cast<ShapedType>();
-  if (!type.hasStaticShape() || type.getSizeInBits() >= fourGbits) {
+  int64_t sizeInBits = type.getNumElements() * type.getElementTypeBitWidth();
```
Comment (krzysz00) [inline mlir/lib/Dialect/Rock/IR/RockDialect.cpp]: I think `getNumElements()` might crash with non-static shapes (do please double-check) and so we might need a two-level if here.

File: external/llvm-project/mlir/include/mlir/Dialect/AMDGPU/IR/AMDGPUDialect.h [hunk 3]
```diff
@@ -14,6 +14,7 @@
 #ifndef MLIR_DIALECT_AMDGPU_IR_AMDGPUDIALECT_H_
 #define MLIR_DIALECT_AMDGPU_IR_AMDGPUDIALECT_H_
 
+#include "mlir/Bytecode/BytecodeOpInterface.h"
```
Comment (krzysz00) [inline external/llvm-project/mlir/include/mlir/Dialect/AMDGPU/IR/AMDGPUDialect.h]: Where's this coming from?

File: external/llvm-project/mlir/lib/Dialect/GPU/IR/GPUDialect.cpp [hunk 4]
```diff
@@ -1395,14 +1383,6 @@ void GPUFuncOp::setWorkgroupAttributionAttr(unsigned index, StringAttr name,
                      getWorkgroupAttribAttrsAttrName());
 }
 
-void GPUFuncOp::setPrivateAttributionAttr(unsigned index, StringAttr name,
```
Comment (krzysz00) [inline external/llvm-project/mlir/lib/Dialect/GPU/IR/GPUDialect.cpp]: Why's this removed? I could've sworn it exists upstream too

**Review by krzysz00 — COMMENTED** [review]
I've got a few minor questions, going off what I've looked at, but I don't think there's any serious issues here

File: external/llvm-project/mlir/lib/Dialect/GPU/IR/GPUDialect.cpp (hunk 4 above)
Comment (giuseros) [reply external/llvm-project/mlir/lib/Dialect/GPU/IR/GPUDialect.cpp]: It does, it is called `setPrivateAttirbutionAttrs` (with an `s` at the end) :)

File: external/llvm-project/mlir/include/mlir/Dialect/AMDGPU/IR/AMDGPUDialect.h (hunk 3 above)
Comment (giuseros) [reply external/llvm-project/mlir/include/mlir/Dialect/AMDGPU/IR/AMDGPUDialect.h]: This is upstream : https://github.com/llvm/llvm-project/blob/main/mlir/include/mlir/Dialect/GPU/IR/GPUDialect.h

File: mlir/lib/Dialect/Rock/IR/RockDialect.cpp (hunk 2 above)
Comment (giuseros) [reply mlir/lib/Dialect/Rock/IR/RockDialect.cpp]: Yep, you are right.

File: external/llvm-project/mlir/lib/Dialect/GPU/IR/GPUDialect.cpp (hunk 4 above)
Comment (krzysz00) [reply external/llvm-project/mlir/lib/Dialect/GPU/IR/GPUDialect.cpp]: ... right, that was me upstreaming this

**Comment by krzysz00** [comment]
Looking at your changes, I'm concerned by the removal of 

```
      IRMapping storeAUpdates, storeBUpdates;

      storeAUpdates.map(blockwiseLoadA.getResult(0),
                        blockwiseLoadAClone.getResult(0));
      storeBUpdates.map(blockwiseLoadB.getResult(0),
                        blockwiseLoadBClone.getResult(0));
      b.clone(*blockwiseStoreA.getOperation(), storeAUpdates);
     b.clone(*blockwiseStoreB.getOperation(), storeBUpdates);
```
in GridwiseGemmToBlockwise . The point of that is that the store is storing the results of the cloned load, but, with the new code, you'll always be using the original load result. I'm not sure that that's correct. Unless I missed a refactoring and it's fine.

**Comment by giuseros** [comment]
> Looking at your changes, I'm concerned by the removal of
> 
> ```
>       IRMapping storeAUpdates, storeBUpdates;
> 
>       storeAUpdates.map(blockwiseLoadA.getResult(0),
>                         blockwiseLoadAClone.getResult(0));
>       storeBUpdates.map(blockwiseLoadB.getResult(0),
>                         blockwiseLoadBClone.getResult(0));
>       b.clone(*blockwiseStoreA.getOperation(), storeAUpdates);
>      b.clone(*blockwiseStoreB.getOperation(), storeBUpdates);
> ```
> 
> in GridwiseGemmToBlockwise . The point of that is that the store is storing the results of the cloned load, but, with the new code, you'll always be using the original load result. I'm not sure that that's correct. Unless I missed a refactoring and it's fine.

So we had a chat about this. When we added the `loadBuffer/storeBuffer` we made the `GlobalLoadLoop` result-less. So that code was segfaulting, and  I had to remove it.

**Comment by krzysz00** [comment]
Update: the code I was worried about was actually wrong this whole time.

In trivialities:

```
diff -rup /home/kdrewnia/llvm-project/mlir/include/mlir/Conversion/TosaToTensor/TosaToTensor.h external/llvm-project/mlir/include/mlir/Conversion/TosaToTensor/TosaToTensor.h
--- /home/kdrewnia/llvm-project/mlir/include/mlir/Conversion/TosaToTensor/TosaToTensor.h        2023-02-22 17:31:02.979976369 +0000
+++ external/llvm-project/mlir/include/mlir/Conversion/TosaToTensor/TosaToTensor.h      2023-02-17 17:47:55.272517815 +0000
@@ -13,6 +13,7 @@
 #ifndef MLIR_CONVERSION_TOSATOTENSOR_TOSATOTENSOR_H
 #define MLIR_CONVERSION_TOSATOTENSOR_TOSATOTENSOR_H

+#include "mlir/IR/PatternMatch.h"
 #include "mlir/Pass/Pass.h"

 namespace mlir {
```
is a stray `#include` of ours we could get rid of if it's not too much bother

So is
```
diff -rup /home/kdrewnia/llvm-project/mlir/include/mlir/Dialect/GPU/IR/GPUDialect.h external/llvm-project/mlir/include/mlir/Dialect/GPU/IR/GPUDialect.h
--- /home/kdrewnia/llvm-project/mlir/include/mlir/Dialect/GPU/IR/GPUDialect.h   2023-07-05 18:39:58.979247149 +0000
+++ external/llvm-project/mlir/include/mlir/Dialect/GPU/IR/GPUDialect.h 2023-07-06 16:42:40.967547851 +0000
@@ -28,6 +28,8 @@
 #include "mlir/Interfaces/SideEffectInterfaces.h"
 #include "llvm/ADT/STLExtras.h"

+#include "llvm/ADT/STLExtras.h"
+
 namespace mlir {
 namespace gpu {
```

And also
```
diff -rup /home/kdrewnia/llvm-project/mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp external/llvm-project/mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp
--- /home/kdrewnia/llvm-project/mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp       2023-07-06 16:42:19.883390745 +0000
+++ external/llvm-project/mlir/lib/Conversion/TosaToLinalg/TosaToLinalg.cpp     2023-07-06 16:42:40.999548088 +0000
@@ -11,6 +11,7 @@
 //===----------------------------------------------------------------------===//

 #include "mlir/Conversion/TosaToLinalg/TosaToLinalg.h"
+#include "mlir/Conversion/TosaToTensor/TosaToTensor.h"
 #include "mlir/Dialect/Arith/IR/Arith.h"
 #include "mlir/Dialect/Arith/Utils/Utils.h"
 #include "mlir/Dialect/Linalg/IR/Linalg.h"
```

We've also got trivial divergences from `mlir/lib/Conversion/TosaToLinalg/TosaToLinalgNamed.cpp ` in the upstream commit, it looks like - can you copy that file in and see if anything breaks?

`GPUFuncOp::setWorkgroupAttributionAttr` is in a different place from upstream - this is a trivial rearrangement but still.

`external/llvm-project/mlir/lib/Dialect/LLVMIR/IR/LLVMDialect.cpp` should be unified with upstream, seeing as inserting symbols uniques their names and their code is simpler.

Why is `mlir/test/Integration/GPU/CUDA/printf.mlir` marked `XFAIL`? It should just not run on our CI and so not be a problem. Could you un-XFAIL it?

**Comment by krzysz00** [comment]
So that's mostly trivialities, but the `XFAIL`'d CUDA printf test is one I'd actually like to see addressed, but that can be a followup ticket.

**Review by krzysz00 — APPROVED** [review]
I've left a pile of random cleanup items, but, overall, having stared at the diffs (the diff of `mlir/` and the `diff -rup` between our `external/llvm-project/mlir` and `mlir/` of the correct upstream commit, I see no real issues.

**Comment by giuseros** [comment]
Hi @krzysz00 , I applied most the small cleanup changes. The unification of the `LLVMDialect.cpp` is something I would leave as a followup

**Comment by krzysz00** [comment]
I have one question: given that there's been MLIR C API functions renamed, do we want to tank the broken MIGraphX build for a few days or do we want someone (if it's me that'll probably need to wait until Monday) to put up a branch that adapts to the changes?

**Comment by giuseros** [comment]
> I have one question: given that there's been MLIR C API functions renamed, do we want to tank the broken MIGraphX build for a few days or do we want someone (if it's me that'll probably need to wait until Monday) to put up a branch that adapts to the changes?

I would stop the MIGraphX build for now, and amend it next week. Maybe @manupak  and @sjw36 might have other suggestions?

**Pull request merged**